### ISSUE TITLE
Idea: maybe logging statements shouldn’t count against line count or ABC size?

### ISSUE THREAD

**Issue by aviflax** [bbatsov/rubocop#3023]
I’m loving RuboCop; especially as someone new to Ruby it’s a fantastic way to incrementally and organically learn many Ruby best practices. Thank you!

I’m struggling now to get a method down below the default thresholds for line count and ABC size, and as an experiment I tried just removing all the logging statements from the method. And that totally worked, removed the warnings. This seems to me sub-optimal because logging statements don’t _actually_ increase the complexity of the method; they’re kinda-sorta incidental.

So: perhaps it should be possible to exclude logging statements from these metrics? Maybe even by default?

Thank you!

**Comment by jdickey** [comment]
I'm having the same problem; I'm just trying to envision how they'd evaluate which are logging lines and which aren't, given the number of different logging packages out there (I use and rave about [Semantic Logger](http://rocketjob.github.io/semantic_logger/), for what that's worth).

If there was a setting that, when enabled, told the method-length and complexity cops not to count lines with the traditional logging-level methods (`%w(trace debug info warn error fatal)`, perhaps), that might be a good start. That wouldn't completely solve the issue for every logger, of course; you still have setup and so on that you do.

In the meantime, when you have methods with logging that get flagged by RuboCop, and removing the logging silences RuboCop, you might want to put disable-/enable-RuboCop comments around the method(s), such as:

```ruby
    # NOTE: RuboCop metric cops for line length and actual complexity are
    #       disabled for `#some_method`; removing the logging calls will bring
    #       the method into compliance; an assertion which should be verified
    #       each time this method is modified.
    # rubocop:disable Metrics/LineLength, Metrics/CyclomaticComplexity
    def some_method
      logger.trace 'Entering #some_method', foo: @foo, bar: @bar
      # ...
      if disaster_strikes!
        logger.error 'Disaster struck!', foo: @foo, other_stuff: other_stuff
        # ...
      end
      # ...
      ret = whatever_computes_your_return_value
      logger.trace 'Leaving #some_method', ret: ret
      ret
    end
    # rubocop:enable Metrics/LineLength, Metrics/CyclomaticComplexity
```

You get the idea. Yes, it's *ugly*, and It Would Be Very Nice if we didn't have to do it. Logging is *important*; many would go as far as to say that if you're not logging, you're flying with at least one eye blindfolded — something that no sane pilot would do. But we don't want to give up RuboCop "just because" we're logging, so we make it *blitheringly* obvious to the next person to read this code what's going on and why.

**Comment by shevegen** [comment]
> In the meantime, when you have methods with logging that get flagged by RuboCop,
> and removing the logging silences RuboCop

This is a bit cumbersome, because you have to add/modify two lines in comments; one for
disabling rubocop, the other for enabling it again.

If one could tag the comment to a method, such as one defined via "def foo", and using a
special keyword in the comment section for rubocop to recognize, then the trailing 
comment part, the ' #rubocop:enable' could be omitted.

It still would be rather cumbersome IMO.

I find line count or size count per line not an extremely useful characteristic altogether. While
I do agree that code should be limited to 80 characters per line, simply due to readability
in how human beings work (easier to capture a smaller area on your screen), I also often
have code that may be indented a bit towards the right, simply due to namespacing alone,
and some longer if checks. If I then use longer variable names too, reaching the 80 lines
limit is quite easy. Maybe the 80 chars should not be a strict default but should be semi-soft
up to 100 or so, but I guess I digress.

**Comment by jdickey** [comment]
@shevegen Oh, I agree with you, for the most part; the only significant difference is that I generally value checks on method length (and IIRC [Reek](https://github.com/troessner/reek)'s defaults are even more stringent than RuboCop's, and dealt with in similar fashion).

I was just pointing out "this is what you can do with any reasonably recent version of RuboCop; it's *fugly*, but it's there, and the ugliness IMO doesn't outweigh the value that RuboCop otherwise brings to the project". No tool is perfect, but any tool whose value exceeds its cost, as RuboCop has for us, is generally seen as a Good Thing.

**Comment by qnm** [comment]
Hey folks,

We at Digivizer have a modified `MethodLength` cop which accepts a blacklist of strings which will excludes a given line from the length count.

The API looks like:

```
DGVZMethodLength:
  Enabled: true
  Max: 23
  Blacklist:
    - .error
    - .info
    - .debug
    - .fatal
    - .trace

```

It works using a simple `include?` check for each term on the blacklist.

If anyone think this is a worthwhile feature, we can put together a PR to get it upstream.

**Comment by aviflax** [comment]
Wow, what a great discussion — I’ve learned so much from this already! Thanks all!

Personally I’m now on the fence WRT this topic, because after I posted this issue I refactored the method I was working on, breaking it into 3 different methods — which both satisfied Rubocop and also made the code easier to read, understand, and test. So in the end, Rubocop’s strictness here helped me out.

**Comment by jdickey** [comment]
@aviflax I've had *numerous* similar experiences — alas, some methods have successfully resisted multiple attempts at such refactoring.

That being said, I'd *love* to see @qnm's `Blacklist` feature submitted as a PR and merged. You may well not be able to see which refactoring makes most sense absent log data, and having RuboCop as part of your commit and/or CI workflows presently makes that harder than it "should" be.

**Comment by aviflax** [comment]
@jdickey I agree on the blacklist feature. You know what else might be cool? The ability to specify a blacklist inline — right in the code — as a special Rubocop “directive” right on a specific method. (Although in that context it might be clearer to use “ignore” rather than “blacklist”.)

**Comment by jdickey** [comment]
@aviflax And, now that you mention it, `Ignore` would be a better config keyword than `Blacklist` (unless it's already dedicated by convention to something else and I just can't remember it at present).

So, @qnm, can this code be contributed as a PR that we can poke at and use? I notice that neither you nor the Digivizer org have a public fork of RuboCop; am I missing something?

**Comment by qnm** [comment]
> So, @qnm, can this code be contributed as a PR that we can poke at and use?

@jdickey Yes, that won't be a problem. We'll put it together tomorrow afternoon.

>  I notice that neither you nor the Digivizer org have a public fork of RuboCop; am I missing something?

Nope - it's a custom cop based on `MethodLength` that currently only exist in our private git repo. I'll sort something out for you tomorrow.

**Comment by jdickey** [comment]
@qnm I'm not a maintainer (yet 😉), but even if it takes a while to massage it into `master`, @aviflax and I would certainly make use of it and possibly propose a couple of riffs of our own. Thanks for this!

**Comment by qnm** [comment]
@jdickey @aviflax I've referenced a PR which contains we're currently using. Hopefully it's also useful to others.

**Comment by jellybob** [comment]
This is awesome - I'd just searched the issues to see if anyone else had discussed something like this. I'm more looking to disable the `LineLength` cop on RSpec lines which are specifying the test (`context`, `describe`, and `it`), but I assume I could build something like that off this.

**Comment by singpolyma** [comment]
Did anyone ever find a solution for this?
